sdk/src/main/scala/com/horizen/account/state/WithdrawalMsgProcessor.scala

```diff
@@ -0,0 +1,195 @@
+    })
+    listOfWithdrawalReqs
+  }
+
+  protected def getListOfWithdrawalReqRecords(msg: Message, view: AccountStateView): Try[ExecutionResult] = {
+    Try {
+      require(msg.getData.length == OP_CODE_LENGTH + Ints.BYTES, s"Wrong data length ${msg.getData.length}")
+      val epochNum = BytesUtils.getInt(msg.getData, OP_CODE_LENGTH)
+
+      val listOfWithdrawalReqs = getListOfWithdrawalReqRecords(epochNum, view)
+
+      val withdrawalRequestSerializer = new ListSerializer[WithdrawalRequest](WithdrawalRequestSerializer)
+      val list = withdrawalRequestSerializer.toBytes(listOfWithdrawalReqs.asJava)
+      new ExecutionSucceeded(gasSpentForGetListOfWithdrawalReqsCmd, list)
+    }
+
+  }
+
+
+
+  private[horizen] def checkWithdrawalRequestValidity(msg: Message, view: AccountStateView): Unit = {
+    val withdrawalAmount = msg.getValue
+
+    if (msg.getData.length != OP_CODE_LENGTH + MCPublicKeyHashProposition.KEY_LENGTH) {
+      log.error(s"Wrong message data field length: ${msg.getData.length}")
+      throw new IllegalArgumentException("Wrong message data field length")
+    }
+    else if (!ZenWeiConverter.isValidZenAmount(withdrawalAmount)) {
+      log.error(s"Withdrawal amount is not a valid Zen amount: $withdrawalAmount")
+      throw new IllegalArgumentException("Withdrawal amount is not a valid Zen amount")
+    }
+    else if (withdrawalAmount.compareTo(DUST_THRESHOLD_IN_WEI) < 0) {
+      log.error(s"Withdrawal amount is under the dust threshold: $withdrawalAmount")
+      throw new IllegalArgumentException("Withdrawal amount is under the dust threshold")
+    }
+    else {
+      val balance = view.getBalance(msg.getFrom.address()).get
+      if (balance.compareTo(withdrawalAmount) < 0) {
+        log.error(s"Insufficient balance amount: balance: $balance, requested withdrawal amount: $withdrawalAmount")
+        throw new IllegalArgumentException("Insufficient balance amount")
+      }
+    }
+
+  }
+
+  protected def addWithdrawalRequest(msg: Message, view: AccountStateView): Try[ExecutionResult] = {
+    Try {
+      checkWithdrawalRequestValidity(msg, view)
+      val currentEpochNum = view.getWithdrawalEpochInfo.epoch
+      val numOfWithdrawalReqs = getWithdrawalEpochCounter(view, currentEpochNum)
+      if (numOfWithdrawalReqs >= MAX_WITHDRAWAL_REQS_NUM_PER_EPOCH) {
+        log.error(s"Reached maximum number of Withdrawal Requests per epoch: request is invalid")
+        new ExecutionFailed(gasSpentForAddNewWithdrawalReqFailure, new IllegalArgumentException("Reached maximum number of Withdrawal Requests per epoch"))
+      }
+
+      val nextNumOfWithdrawalReqs: Int = numOfWithdrawalReqs + 1
+      val key = getWithdrawalEpochCounterKey(currentEpochNum)
+      view.updateAccountStorageBytes(fakeSmartContractAddress.address(), key, Ints.toByteArray(nextNumOfWithdrawalReqs)).get
+
+      val mcDestAddr = msg.getData.slice(OP_CODE_LENGTH, OP_CODE_LENGTH + MCPublicKeyHashProposition.KEY_LENGTH)
```

Comment:
do we check if there are any trash bytes after?

Reply:
In  checkWithdrawalRequestValidity the data length is verified that is exactly  OP_CODE_LENGTH + MCPublicKeyHashProposition.KEY_LENGTH